### PR TITLE
Wrapper for PGoApi and PGoApiRequest, implements auto-retrying on API request failures.

### PR DESCRIPTION
**Pull request by sebastienvercammen**
## Description

* Added `PGoApiWrapper` that wraps around a `PGoApi` object and overwrites its `.create_request` method to return a wrapped `PGoApiRequest`.
* Added `PGoRequestWrapper` to wrap `PGoApiRequest`, implementing auto-retrying on API request failure.
* Updated webhook docs because people on Discord need to know "pokemon" is also a type.

## Motivation and Context
Consistent behavior.

## How Has This Been Tested?
Tested with @Alderon86.

## Screenshots (if appropriate):
```
2017-07-27 22:47:33,625 [   search-worker-0][pgorequestwrapper][   DEBUG] Wrapped PGoApiRequest.
2017-07-27 22:47:33,625 [   search-worker-0][pgorequestwrapper][   DEBUG] Sending API request. Retries left: 3.
2017-07-27 22:47:33,795 [   search-worker-0][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-27 22:47:33,795 [   search-worker-0][pgorequestwrapper][   DEBUG] Sending API request. Retries left: 2.
2017-07-27 22:47:33,825 [   search-worker-0][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-27 22:47:33,825 [   search-worker-0][pgorequestwrapper][   DEBUG] Sending API request. Retries left: 1.
2017-07-27 22:47:33,854 [   search-worker-0][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-27 22:47:33,854 [   search-worker-0][pgorequestwrapper][   DEBUG] Sending API request. Retries left: 0.
2017-07-27 22:47:33,891 [   search-worker-0][       account][   ERROR] Login for account -censored- failed. Exception in call request: Unexpected HTTP server response - needs 200 got 404. .
```

## Types of changes
- [x] Enhancement

## Checklist:
- [x] My code follows the code style of this project.

### CODE REVIEW

**Comment by friscoMad** [comment]
This are the only exceptions that make sense to retry, not sure if all those are in use currently:

class HashServerException(PgoapiError):
class TimeoutException(PgoapiError):
class AuthTimeoutException(AuthException, TimeoutException):
class BadRequestException(PgoapiError):
class BadHashRequestException(BadRequestException):
class MalformedNianticResponseException(PgoapiError):
class MalformedHashResponseException(MalformedResponseException, HashServerException):
class ServerBusyOrOfflineException(PgoapiError):
class NianticOfflineException(ServerBusyOrOfflineException):
class NianticTimeoutException(NianticOfflineException, TimeoutException):
class HashingOfflineException(ServerBusyOrOfflineException, HashServerException):
class HashingTimeoutException(HashingOfflineException, TimeoutException):
class ServerSideRequestThrottlingException(PgoapiError):
class NianticThrottlingException(ServerSideRequestThrottlingException):
class UnexpectedResponseException(PgoapiError):
class UnexpectedHashResponseException(UnexpectedResponseException, HashServerException):
class NianticIPBannedException(ServerSideAccessForbiddenException):

Also it would be great if all niantic related exception would use resend the send message, without recreating the whole message (cpu and hash consumption)

File: pogom/pgorequestwrapper.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        except:
+            pass
+
+        # If we've reached here, an exception was raised and the function
+        # hasn't returned yet.
+
+        # Try again if we have retries left.
+        if retries_left > 0:
+            # Rotate proxy.
+            proxy_idx, proxy_url = get_new_proxy(get_args())
```
Comment (friscoMad) [inline pogom/pgorequestwrapper.py]: it needs to be under args.proxy check

**Comment by sebastienvercammen** [comment]
@friscoMad Implementing all specific exceptions for a list that huge is unnecessary: we've had months to see what errors take up 99% of the exceptions that happen for our users and in practice it's only a handful. However, all other exception types still exist and should still be retried. Your list is also missing all `requests` and `urllib` related exceptions.

Keeping it as is makes sense - any exceptions we catch will be related to our API request, so a retry will never hurt, and we'll be sure that we're not missing any exceptions (now or in the future).

**Comment by nhumber** [comment]
working with latest changes.
wrapper handles retrying on fails

```
2017-07-28 07:04:14,862 [  search-worker-30][pgorequestwrapper][   DEBUG] Wrapped PGoApiRequest.
2017-07-28 07:04:14,866 [  search-worker-30][pgorequestwrapper][   DEBUG] Sending API request w/ 3 retries left.
2017-07-28 07:04:14,869 [  search-worker-30][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-28 07:04:14,870 [  search-worker-30][pgorequestwrapper][   DEBUG] Sending API request w/ 2 retries left.
2017-07-28 07:04:14,873 [  search-worker-30][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-28 07:04:14,873 [  search-worker-30][pgorequestwrapper][   DEBUG] Sending API request w/ 1 retries left.
2017-07-28 07:04:14,876 [  search-worker-30][pgorequestwrapper][   DEBUG] API request failed. Retrying...
2017-07-28 07:04:14,877 [  search-worker-30][pgorequestwrapper][   DEBUG] Sending API request w/ 0 retries left.
```

**Pull request merged**